### PR TITLE
[OSGi] Support the Data Service Specification for JDBC™ Technology

### PR DESCRIPTION
**Pull request by laeubi**
Currently users are enforced to write their own code or using third party wrappers to use MariaDB together with OSGi JDBC/JPA Service,

This PR adds support for the [OSGi Data Service Specification for JDBC™ Technology](https://osgi.org/specification/osgi.cmpn/7.0.0/service.jdbc.html)
This allows users of OSGi to use the MariaDB driver in a generic fashion. This is also a perquisite for using MariaDB together with the OSGi JPA Specification.

Since this part of the code is only activated inside an OSGi-Framework for all other users this change is completely transparent.

### CODE REVIEW

File: src/main/java/org/mariadb/jdbc/internal/osgi/MariaDbDataSourceFactory.java [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+
+import org.mariadb.jdbc.Driver;
+import org.mariadb.jdbc.MariaDbDataSource;
+import org.osgi.service.jdbc.DataSourceFactory;
+
+public class MariaDbDataSourceFactory implements DataSourceFactory {
+
+  @Override
+  public MariaDbDataSource createDataSource(Properties props) throws SQLException {
+    MariaDbDataSource source = new MariaDbDataSource();
```
Comment (rusher) [inline src/main/java/org/mariadb/jdbc/internal/osgi/MariaDbDataSourceFactory.java]: Implementation must distinguished basic implementation (MariaDbDataSource) and Connection pooling implementation (MariaDbPoolDataSource) if according properties JDBC_INITIAL_POOL_SIZE, JDBC_MAX_IDLE_TIME, JDBC_MAX_POOL_SIZE, JDBC_MAX_STATEMENTS, JDBC_MIN_POOL_SIZE are set

JDBC_PROPERTY_CYCLE

File: src/main/java/org/mariadb/jdbc/internal/osgi/MariaDbActivator.java [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+public class MariaDbActivator implements BundleActivator {
+
+  private ServiceRegistration<DataSourceFactory> service;
+
+  @Override
+  public void start(BundleContext context) throws Exception {
+    Dictionary<String, Object> properties = new Hashtable<>();
+    properties.put(DataSourceFactory.OSGI_JDBC_DRIVER_CLASS, Driver.class.getName());
+    properties.put(DataSourceFactory.OSGI_JDBC_DRIVER_NAME, "JDBC driver for MariaDB and MySQL");
+    properties.put(DataSourceFactory.OSGI_JDBC_DRIVER_VERSION, Version.majorVersion + "." + Version.minorVersion);
```
Comment (rusher) [inline src/main/java/org/mariadb/jdbc/internal/osgi/MariaDbActivator.java]: Version.version would be better (like '2.4.1-SNAPSHOT', compare to '2.4' )

File: src/main/java/org/mariadb/jdbc/internal/osgi/MariaDbActivator.java [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+ */
+public class MariaDbActivator implements BundleActivator {
+
+  private ServiceRegistration<DataSourceFactory> service;
+
+  @Override
+  public void start(BundleContext context) throws Exception {
+    Dictionary<String, Object> properties = new Hashtable<>();
+    properties.put(DataSourceFactory.OSGI_JDBC_DRIVER_CLASS, Driver.class.getName());
+    properties.put(DataSourceFactory.OSGI_JDBC_DRIVER_NAME, "JDBC driver for MariaDB and MySQL");
```
Comment (rusher) [inline src/main/java/org/mariadb/jdbc/internal/osgi/MariaDbActivator.java]: This correspond to MariaDbDatabaseMetaData.DRIVER_NAME

File: src/main/java/org/mariadb/jdbc/internal/osgi/MariaDbDataSourceFactory.java (hunk 1 above)
Comment (laeubi) [reply src/main/java/org/mariadb/jdbc/internal/osgi/MariaDbDataSourceFactory.java]: One Problem I noticed: MariaDbPoolDataSource does not implements ConnectionPoolDataSource but DataSourceFactory#createConnectionPoolDataSource needs one.
On the other hand MariaDbDataSource DOES implement ConnectionPoolDataSource even though it seems it is not really one.

File: src/main/java/org/mariadb/jdbc/internal/osgi/MariaDbActivator.java (hunk 3 above)
Comment (laeubi) [reply src/main/java/org/mariadb/jdbc/internal/osgi/MariaDbActivator.java]: Thanks for the hint, I have changed this

File: src/main/java/org/mariadb/jdbc/internal/osgi/MariaDbActivator.java (hunk 2 above)
Comment (laeubi) [reply src/main/java/org/mariadb/jdbc/internal/osgi/MariaDbActivator.java]: Changed to Version.version

File: src/main/java/org/mariadb/jdbc/internal/osgi/MariaDbDataSourceFactory.java (hunk 1 above)
Comment (laeubi) [reply src/main/java/org/mariadb/jdbc/internal/osgi/MariaDbDataSourceFactory.java]: Maybe it was meant that MariaDbPoolDataSource extends MariaDbDataSource and implements ConnectionPoolDataSource?

Comment (rusher) [reply src/main/java/org/mariadb/jdbc/internal/osgi/MariaDbDataSourceFactory.java]: Right, that has to be corrected in 2.5. Issue created: https://jira.mariadb.org/browse/CONJ-692

Comment (laeubi) [reply src/main/java/org/mariadb/jdbc/internal/osgi/MariaDbDataSourceFactory.java]: So would it be possible to use the implementation as-is (even if not 100% correct in the sense of connection pooling) and change it when CONJ-692 is fixed?

**Comment by laeubi** [comment]
How can we proceed on this?

**Comment by rusher** [comment]
@laeubi 

Hi, remains the point commented before and similar to other open source projects, the MariaDB Foundation needs to have shared ownership of all code that is included in the MariaDB distribution. The easiest way to achieve this is by submitting your code under the BSD-new license. (The other alternative is to sign the code contribution agreement which can be found here: https://mariadb.com/kb/en/mariadb/mca/)

Please indicate in a comment below that you are contributing your new code of the whole pull request under the BSD-new license or that you have filled out the contribution agreement and sent it.

Thanks,
diego

**Comment by laeubi** [comment]
I'm contributing the new code of the whole pull request under the BSD-new license, is there any template I need to add to the changed file?

File: pom.xml [hunk 4]
```diff
@@ -294,9 +295,8 @@
                             <Bundle-SymbolicName>org.mariadb.jdbc</Bundle-SymbolicName>
                             <Automatic-Module-Name>org.mariadb.jdbc</Automatic-Module-Name>
                             <Export-Package>org.mariadb.jdbc</Export-Package>
-                            <Import-Package>
-                                javax.naming,javax.management,javax.net;resolution:=optional,javax.net.ssl;resolution:=optional,javax.sql,javax.transaction.xa;resolution:=optional,org.slf4j;resolution:=optional
-                            </Import-Package>
+                            <Import-Package>org.osgi.framework,javax.naming,javax.management,javax.net;resolution:=optional,javax.net.ssl;resolution:=optional,javax.sql,org.osgi.service.jdbc;resolution:=optional,javax.transaction.xa;resolution:=optional,org.slf4j;resolution:=optional</Import-Package>
```
Comment (jakefeasel) [inline pom.xml:298]: I think you missed javax.security in this list. It prevents the driver from making SSL connections from an OSGi environment.

Comment (laeubi) [reply pom.xml:298]: I just added what was required for the jdbc service, if you think more packages are required I think you can simply open a new PR as this one is already closed.

**Comment by laeubi** [comment]
@rusher I just noticed that it seems the code for supporting this OSGi specification was removed from maria-db is there a specific reason? Could it be added back again?

**Pull request merged**